pkg/operators/otel-profiles/otel-profiles.go

```diff
@@ -0,0 +1,429 @@
+	"fmt"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"go.opentelemetry.io/collector/pdata/pprofile"
+	"go.opentelemetry.io/collector/pdata/pprofile/pprofileotlp"
+	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/encoding/gzip"
+
+	"github.com/inspektor-gadget/inspektor-gadget/internal/version"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/config"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	apihelpers "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api-helpers"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/otel-profiles/orderedset"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	ParamOtelProfilesExporter = "otel-profiles-exporter"
+
+	ExporterOTLPGRPC = "otlp-grpc"
+
+	CompressionNone = "none"
+	CompressionGZIP = "gzip"
+
+	stackFieldAnnotation      = "profiles.stack-field"
+	valueFieldAnnotation      = "profiles.value-field"
+	sampleAttributeAnnotation = "profiles.sample-attribute"
+)
+
+var supportedExporters = []string{ExporterOTLPGRPC}
+
+type profileConfig struct {
+	Exporter    string `json:"exporter" yaml:"exporter"`
+	Endpoint    string `json:"endpoint" yaml:"endpoint"`
+	Insecure    bool   `json:"insecure" yaml:"insecure"`
+	Compression string `json:"compression" yaml:"compression"`
+}
+
+type otelProfilesOperator struct {
+	clients  map[string]pprofileotlp.GRPCClient
+	callOpts []grpc.CallOption
+}
+
+func (o *otelProfilesOperator) Name() string {
+	return "otel-profiles"
+}
+
+func (o *otelProfilesOperator) Init(params *params.Params) error {
+	if config.Config == nil {
+		return nil
+	}
+
+	o.clients = make(map[string]pprofileotlp.GRPCClient)
+	o.callOpts = make([]grpc.CallOption, 0)
+
+	configs := make(map[string]*profileConfig, 0)
+	err := config.Config.UnmarshalKey("operator.otel-profiles.exporters", &configs)
+	if err != nil {
+		log.Warnf("failed to load operator.otel-profiles.exporters: %v", err)
+	}
+
+	for k, v := range configs {
+		if v.Exporter != ExporterOTLPGRPC {
+			return fmt.Errorf("unsupported log exporter %q; expected one of %s", v.Exporter,
+				strings.Join(supportedExporters, ", "))
+		}
+		var options []grpc.DialOption
+
+		if v.Insecure {
+			options = append(options, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		}
+		switch v.Compression {
+		default:
+			return fmt.Errorf("unsupported log compression %q", v.Compression)
+		case "", CompressionNone:
+		case CompressionGZIP:
+			o.callOpts = append(o.callOpts, grpc.UseCompressor(gzip.Name))
+		}
+
+		conn, err := grpc.NewClient(v.Endpoint, options...)
+		if err != nil {
+			log.Errorf("creating gRPC connection to %q: %v", v.Endpoint, err)
+			return fmt.Errorf("creating gRPC connection to %q: %w", v.Endpoint, err)
+		}
```

Comment:
If this is done in `Init()`, does it mean it will connect during `GetGadgetInfo()` or `ig run trace_exec --help`?

It looks like it should be postponed to `InstantiateDataOperator()`.

Reply:
Unfortunately `InstantiateDataOperator` doesn't work as it'll create a connection per gadget, and we want gadgets to reuse the same connector. We could use a lazy approach to initialize those connections but that's too complex IMO. I'd say we can live with this inefficiency for now.